ttnn/cpp/ttnn/operations/ccl/ccl_op_fusion.cpp

```diff
@@ -0,0 +1,179 @@
+
+    rt_args.push_back(static_cast<uint32_t>(wait_for_start_signal));
+    rt_args.push_back(static_cast<uint32_t>(send_start_signal));
+
+    if (send_start_signal) {
+        rt_args.push_back(static_cast<uint32_t>(start_signal_core_sem_pair->core.x));
+        rt_args.push_back(static_cast<uint32_t>(start_signal_core_sem_pair->core.y));
+        rt_args.push_back(static_cast<uint32_t>(start_signal_core_sem_pair->sem_id));
+    }
+
+}
+
+
+// Used to propagate semaphore information from matmul to all_gather in all_gather_matmul op
+void MatmulFusedOpSignaler::init_all_gather(
+    uint32_t num_transfers,
+    uint32_t ring_size,
+    uint32_t start_ring_index,
+    uint32_t tensor_slice_shape_width,
+    uint32_t output_page_offset,
+    bool is_clockwise_direction,
+
+    uint32_t weight_output_page_offset
+) {
+    this->num_transfers = num_transfers;
+    this->ring_size = ring_size;
+    this->start_ring_index = start_ring_index;
+    this->tensor_slice_shape_width = tensor_slice_shape_width;
+    this->output_page_offset = output_page_offset;
+    this->is_clockwise_dir = is_clockwise_direction;
+
+    this->weight_output_page_offset = weight_output_page_offset;
+
+    initialized_all_gather = true;
+}
+
+void MatmulFusedOpSignaler::init_fused_op(
+    Program& program,
+    Device const* device,
+    const std::variant<CoreRange, CoreRangeSet>& core_range_to_signal
+) {
+    // Clear the existing receiver cores
+    this->fused_op_receiver_cores_noc.clear();
+
+    // Visit the variant to handle CoreRange and CoreRangeSet differently
+    std::visit([&](auto& arg) {
+        using T = std::decay_t<decltype(arg)>;
+        if constexpr (std::is_same_v<T, CoreRange>) {
+            // Handle CoreRange
+            const auto& cores = grid_to_cores(arg.start_coord, arg.end_coord, true);
+            for (auto& core : cores) {
+                this->fused_op_receiver_cores_noc.push_back(device->worker_core_from_logical_core(core));
+            }
+        } else if constexpr (std::is_same_v<T, CoreRangeSet>) {
+            // Handle CoreRangeSet
+            for (const auto& range : arg.ranges()) {
+                const auto& cores = grid_to_cores(range.start_coord, range.end_coord, true);
+                for (auto& core : cores) {
+                    this->fused_op_receiver_cores_noc.push_back(device->worker_core_from_logical_core(core));
+                }
+            }
+        }
+    }, core_range_to_signal);
+
+    // Create the semaphores
+    this->fused_op_receiver_signal_semaphores.push_back(CreateSemaphore(program, core_range_to_signal, 0));
+    this->fused_op_receiver_signal_semaphores.push_back(CreateSemaphore(program, core_range_to_signal, 0));
+
+    // Set the number of fused op cores to signal
+    this->num_fused_op_cores_to_signal = this->fused_op_receiver_cores_noc.size();
+
+    initialized_fused_op = true;
+}
+
+void MatmulFusedOpSignaler::emit_matmul_fused_op_rt_args(
+    std::vector<uint32_t>& rt_args,
```

Comment:
same as above.
But looks like we have similar code everywhere. So it is up to you if you want to fix the style here.